File: lib/ts/src/internal/schema/ParseHeadersType.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { TelepactSchemaParseError } from '../../TelepactSchemaParseError.js';
import { getTypeUnexpectedParseFailure } from '../../internal/schema/GetTypeUnexpectedParseFailure.js';
import { TFieldDeclaration } from '../types/TFieldDeclaration.js';
import { ParseContext } from '../../internal/schema/ParseContext.js';
import { parseStructFields } from './ParseStructFields.js';
import { SchemaParseFailure } from './SchemaParseFailure.js';
import { THeaders } from '../types/THeaders.js';

export function parseHeadersType(
    path: any[],
    headersDefinitionAsParsedJson: { [key: string]: any },
    schemaKey: string,
    ctx: ParseContext,
): THeaders {
    const parseFailures: SchemaParseFailure[] = [];
    const requestHeaders: { [key: string]: TFieldDeclaration } = {};
    const responseHeaders: { [key: string]: TFieldDeclaration } = {};

    const requestHeadersDef = headersDefinitionAsParsedJson[schemaKey];

    const thisPath = [...path, schemaKey];

    if (
        typeof requestHeadersDef !== 'object' ||
        Array.isArray(requestHeadersDef) ||
        requestHeadersDef === null ||
        requestHeadersDef === undefined
    ) {
        const branchParseFailures = getTypeUnexpectedParseFailure(
            ctx.documentName,
            thisPath,
            requestHeadersDef,
            'Object',
        );
        parseFailures.push(...branchParseFailures);
    } else {
        try {
            const isHeader = true;
            const requestFields = parseStructFields(thisPath, requestHeadersDef, isHeader, ctx);

            // All headers are optional
            for (const field in requestFields) {
                requestFields[field].optional = true;
            }

            Object.assign(requestHeaders, requestFields);
        } catch (e) {
            if (e instanceof TelepactSchemaParseError) {
                parseFailures.push(...e.schemaParseFailures);
            } else {
                throw e;
            }
        }
    }

    const responseKey = '->';
    const responsePath = [...path, responseKey];

    if (!(responseKey in headersDefinitionAsParsedJson)) {
        parseFailures.push(
            new SchemaParseFailure(ctx.documentName, responsePath, 'RequiredObjectKeyMissing', {
                key: responseKey,
            }),
        );
    }

    const responseHeadersDef = headersDefinitionAsParsedJson[responseKey];

    if (
        typeof responseHeadersDef !== 'object' ||
        Array.isArray(responseHeadersDef) ||
        responseHeadersDef === null ||
        responseHeadersDef === undefined
    ) {
        const branchParseFailures = getTypeUnexpectedParseFailure(
            ctx.documentName,
            thisPath,
            responseHeadersDef,
            'Object',
        );
        parseFailures.push(...branchParseFailures);
    } else {
        try {
            const isHeader = true;
            const responseFields = parseStructFields(responsePath, responseHeadersDef, isHeader, ctx);

            // All headers are optional
            for (const field in responseFields) {
                responseFields[field].optional = true;
            }

            Object.assign(responseHeaders, responseFields);
        } catch (e) {
            if (e instanceof TelepactSchemaParseError) {
                parseFailures.push(...e.schemaParseFailures);
            } else {
                throw e;
            }
        }
    }

    if (parseFailures.length > 0) {
        throw new TelepactSchemaParseError(parseFailures, ctx.telepactSchemaDocumentNamesToJson);
    }

    return new THeaders(schemaKey, requestHeaders, responseHeaders);
}

File: lib/go/internal/schema/ParseHeadersType.go
//|
//|  Copyright The Telepact Authors
//|
//|  Licensed under the Apache License, Version 2.0 (the "License");
//|  you may not use this file except in compliance with the License.
//|  You may obtain a copy of the License at
//|
//|  https://www.apache.org/licenses/LICENSE-2.0
//|
//|  Unless required by applicable law or agreed to in writing, software
//|  distributed under the License is distributed on an "AS IS" BASIS,
//|  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//|  See the License for the specific language governing permissions and
//|  limitations under the License.
//|

package schema

import "github.com/telepact/telepact/lib/go/internal/types"

// ParseHeadersType parses a headers definition into a THeaders value.
func ParseHeadersType(path []any, headersDefinition map[string]any, schemaKey string, ctx *ParseContext) (*types.THeaders, error) {
	if ctx == nil {
		return nil, nil
	}

	parseFailures := make([]*SchemaParseFailure, 0)
	requestHeaders := make(map[string]*types.TFieldDeclaration)
	responseHeaders := make(map[string]*types.TFieldDeclaration)

	thisPath := append(append([]any{}, path...), schemaKey)
	requestHeadersDef, hasRequest := headersDefinition[schemaKey]
	if !hasRequest {
		requestHeadersDef = nil
	}

	requestHeadersMap, ok := requestHeadersDef.(map[string]any)
	if !ok {
		parseFailures = append(parseFailures, GetTypeUnexpectedParseFailure(ctx.DocumentName, thisPath, requestHeadersDef, "Object")...)
	} else {
		fields, err := ParseStructFields(thisPath, requestHeadersMap, true, ctx)
		if err != nil {
			if parseErr, ok := err.(*ParseError); ok {
				parseFailures = append(parseFailures, parseErr.Failures...)
			} else {
				return nil, err
			}
		} else {
			for key, field := range fields {
				if field != nil {
					field.Optional = true
				}
				requestHeaders[key] = field
			}
		}
	}

	responseKey := "->"
	responsePath := append(append([]any{}, path...), responseKey)

	responseHeadersDef, hasResponse := headersDefinition[responseKey]
	if !hasResponse {
		parseFailures = append(parseFailures, NewSchemaParseFailure(ctx.DocumentName, responsePath, "RequiredObjectKeyMissing", map[string]any{"key": responseKey}))
	}

	responseHeadersMap, ok := responseHeadersDef.(map[string]any)
	if !ok {
		if hasResponse {
			parseFailures = append(parseFailures, GetTypeUnexpectedParseFailure(ctx.DocumentName, responsePath, responseHeadersDef, "Object")...)
		}
	} else {
		fields, err := ParseStructFields(responsePath, responseHeadersMap, true, ctx)
		if err != nil {
			if parseErr, ok := err.(*ParseError); ok {
				parseFailures = append(parseFailures, parseErr.Failures...)
			} else {
				return nil, err
			}
		} else {
			for key, field := range fields {
				if field != nil {
					field.Optional = true
				}
				responseHeaders[key] = field
			}
		}
	}

	if len(parseFailures) > 0 {
		return nil, &ParseError{Failures: parseFailures, DocumentJSON: ctx.TelepactSchemaDocumentNamesToJSON}
	}

	return types.NewTHeaders(schemaKey, requestHeaders, responseHeaders), nil
}
